design-proposals/garbage-collect-completed-dvs.md

```diff
@@ -0,0 +1,60 @@
+(who is this feature set intended for)
+
+## User Stories
+* As a VM owner I use a standalone DataVolume to prepare a VM disk PVC for use by my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. The PVC should still maintain an independent lifecycle from the VM.
+* As a VM owner I add an entry to the dataVolumeTemplates section of my VM spec to have a disk prepared for my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. I want kubevirt to understand that it doesn't need to create a new DataVolume when the underlying PVC exists. The PVC should still share the same lifecycle as the VM.
+* As a VM owner I want to restore my backed up VM without needing to recreate the DataVolume.
+
+## Repos
+* **CDI**: Controllers (DataVolume/Import/Upload/Clone/DataImportCron/DataSource) will need adaptations:
+** So far the controllers assumed the existance of DV-PVC pairs, which is no longer a valid assumption.
+** DataImportCron controller currently re-creates the import DV if deleted, even if the PVC was already populated.
+* **KubeVirt**: VM controller adaptations
+# Design
+## PVC Ownership
+The DataVolume controller creates a PVC owned by its DV. When the PVC is populated:
+* PVC ownership by the DV will be removed
+* If DV has ownership (e.g. by VM, when created via DataVolumeTemplates), it will be transferred to its PVC
+* Any DV information that might be useful after it is garbage collected, should be annotated on the PVC (TBD)
+
+If the PVC population failed, the DV should remain the owner of the PVC.
+
+## Garbage Collection
+When DataVolume does not own a PVC, it can be deleted immediately, but we may want to keep it for a while:
+* Add CDI config minimal time for DV garbage collection after completion (disabled by default, for opt-in)
+  `dataVolumeTTL *int32`
+* Add DV annotation to prevent its garbage collection
+  `"cdi.kubevirt.io/storage.dv.retainAfterCompletion": "true"`
+
+Garbage collection will be executed whenever a DV phase is updated to Succeeded
+
+## API Examples
+(tangible API examples used for discussion)
+
+## Scalability
+(overview of how the design scales)
+
+## Update/Rollback Compatibility
+ Garbage collection opt-in is introduced. We change the fundamental expectations that exist for how DataVolumes behave now, and a change in default behavior is a breakage in compatibility. Therefore, existing KubeVirt + CDI combinations will continue to work even after updating CDI, and new/updated installs can opt-in to the garbage collection behavior if they'd like. This opt-in will be a global setting on the CDI Config where the garbage collection TTL is configured (disabled by default). That would solve the OpenShift Virtualization use case where we have tight control over the KubeVirt + CDI combinations. We should handle a few odd conditions as a result of this change that could impact compatibility.
```

Comment:
I am not sure what this means

"That would solve the OpenShift Virtualization use case where we have tight control over the KubeVirt + CDI combinations."

Reply:
I think it means that in the downstream case we have more freedom to decide which kubevirt version is bundled with which CDI, so we can `prevent someone from pairing a new version of CDI (which automatically garbage collects) with an old version of KubeVirt (which recreates after garbage collection)` or backport logic that handles it to respective versions

Reply:
@akalenyu exactly.